### PR TITLE
Get rid of the fake stack frame for reading from constants

### PR DESCRIPTION
**Pull request by oli-obk**
r? @RalfJung 

fixes the ice in https://github.com/rust-lang/rust/issues/53708 but still keeps around the wrong "non-exhaustive match" error

cc @varkor

### CODE REVIEW

**Comment by RalfJung** [comment]
(I'm on a conference, and will start catching up with my emails on Jan 22nd.)

**Comment by oli-obk** [comment]
@rust-timer build a5a18a4e3822d2c19cb9cc6752bcc8930f8162e8

File: src/librustc_mir/interpret/eval_context.rs [hunk 1]
```diff
@@ -274,6 +280,20 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tc
         }
     }
 
+    pub fn monomorphize_in_frame<T: TypeFoldable<'tcx> + Subst<'tcx>>(
```
Comment (RalfJung) [inline src/librustc_mir/interpret/eval_context.rs]: This name is odd, we don't have anything else called like it.  `monomorphize` would match better the fact that `resolve` and `subst_and_normalize_erasing_regions` also just implicitly depend on the frame.

Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs]: I swapped the naming scheme, so `monomorphize` is the one with only a type argument and `monomorphize_with_substs` doesn't need a stack frame to get the substs from.

File: src/librustc_mir/interpret/eval_context.rs [hunk 2]
```diff
@@ -274,6 +280,20 @@ impl<'a, 'mir, 'tcx: 'mir, M: Machine<'a, 'mir, 'tcx>> EvalContext<'a, 'mir, 'tc
...
+            Some(frame) => Ok(self.monomorphize(t, frame.instance.substs)),
+            None => if t.needs_subst() {
+                err!(TooGeneric).into()
+            } else {
+                Ok(t)
+            },
+        }
+    }
+
     pub fn monomorphize<T: TypeFoldable<'tcx> + Subst<'tcx>>(
```
[the remark this answers is not in the document]
Comment (oli-obk) [reply src/librustc_mir/interpret/eval_context.rs]: done

**Comment by RalfJung** [comment]
Thanks!

@bors r+

**Pull request merged**